File: Algorithms/C++/00225-Implement Stack using Queues.cpp
class MyStack {
public:
    std::queue<int> q1;
    std::queue<int> q2;
    bool queueOne = true;

    MyStack() {
        
    }
    
    void push(int x) {
        int front{0};
        if(queueOne){
            q2.push(x);
            while(!q1.empty()){
                front = q1.front();
                q2.push(front);
                q1.pop();
            }
            queueOne = false;
        } else{
            q1.push(x);
            while(!q2.empty()){
                front = q2.front();
                q1.push(front);
                q2.pop();
            }
            queueOne = true;
        }
    }
    
    int pop() {
        int front{0};
        if(queueOne){
            front = q1.front();
            q1.pop();
        } else {
            front = q2.front();
            q2.pop();         
        }
        return front;
    }
    
    int top() {
        return queueOne ? q1.front() : q2.front();
    }
    
    bool empty() {
        return queueOne ? q1.empty() : q2.empty();
    }

};

/*
    tc: O(n)

    [1,2,3,4]

    []
    [4,3,2,1]

*/


/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack* obj = new MyStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * bool param_4 = obj->empty();
 */

File: Algorithms/Python3/00225-Implement Stack using Queues.py
class MyStack:
    def __init__(self):
        self.q = []    # queue to store stack values

    def push(self, x: int) -> None:
        self.q.append(x)

    def pop(self) -> int:
        size = len(self.q)
        for i in range(size-1):
            self.q.append(self.q[i])

        for i in range(size-1):
            self.q.pop(0)
        
        value = self.q[0]
        self.q.pop(0)
        return value
        
    def top(self) -> int:
        return self.q[-1]

    def empty(self) -> bool:
        return len(self.q) == 0
        
'''
    pop()
    [1, 2, 3]
    [1, 2, 3, 1, 2]
    [3, 1, 2]
    [1, 2]
    return 3
    
    can only use 
        list.append()
        list[0]
        len(list)
        list
'''

# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()